[Origo] Adds support to mark storage-node/distrubtion-node under maintenance

zeeshanakram3 opened the pull request:
Addresses #4765 

- [x]  Adds Storage/Distribution QN mappings
- [x] Adds `operator:set-metadata` & `leader:set-node-operational-status` CLI commands
- [x] ~WIP: Updating Storage/Distribution related integration tests~ https://github.com/Joystream/joystream/issues/4815

zeeshanakram3 committed:
add protobuf schemas for operational status feat

zeeshanakram3 committed:
add QN schemas for operational status feat

zeeshanakram3 committed:
add QN mappings for operational status feat.

zeeshanakram3 committed:
add CLI commands to set operational status in Argus

zeeshanakram3 committed:
add CLI commands to set operational status in Colossus

zeeshanakram3 committed:
argus: generate-docs

zeeshanakram3 committed:
Merge remote-tracking branch 'upstream/master' into origo_argus_colossus_operational_status_feat

mnaamani commented:
You need to add implicitly new dependency on inquirer-datepicker? (and maybe others) to storage node: this is why this check is failing building docker image for storage node https://github.com/Joystream/joystream/actions/runs/5561050023/jobs/10158465896?pr=4793#step:8:373

zeeshanakram3 committed:
add inquirer dependency in storage-node package.json

zeeshanakram3 commented:
Integration tests will be added in #4815

mnaamani commented:
@zeeshan perhaps we can try to bump version of `protobufjs` in this PR?

see https://github.com/advisories/GHSA-h755-8qp9-cq85

zeeshanakram3 committed:
Merge remote-tracking branch 'upstream/master' into origo_argus_colossus_operational_status_feat

zeeshanakram3 committed:
bump package versions & update changelog

zeeshanakram3 commented:
> perhaps we can try to bump version of `protobufjs` in this PR?

Bumped package versions and updated CHANGELOG.md in [d133ffb](https://github.com/Joystream/joystream/pull/4793/commits/d133ffbf1f3265f560d5aaece632820c08c21e0f)

mnaamani commented:
> > perhaps we can try to bump version of `protobufjs` in this PR?
> 
> Bumped package versions and updated CHANGELOG.md in [d133ffb](https://github.com/Joystream/joystream/pull/4793/commits/d133ffbf1f3265f560d5aaece632820c08c21e0f)

Great.
I think QN package also needs a version bump.

zeeshanakram3 committed:
bump QN package version

zeeshanakram3 commented:
> I think QN package also needs a version bump.

Done in [ba24cfe](https://github.com/Joystream/joystream/pull/4793/commits/ba24cfe683ee6c29f3f5ba43c9152d4736a4f50a)

kdembler commented on metadata-protobuf/proto/Storage.proto:
```diff
@@ -46,3 +48,31 @@ message DistributionBucketFamilyMetadata {
+  }
+
+  // Node's Operational status to set
+  optional OperationalStatus status = 1; 
```
Shouldn't the `status` be required here? Full `NodeOperationalStatusMetadata` is already optional in `DistributionBucketOperatorMetadata`

kdembler replied:
@zeeshanakram3 still relevant

mnaamani replied:
I would go further and suggest that we should probably have multiple messages, corresponding with the options in the commands for setting the metadata.

```
message NodeOperationalStatusNormal {
  optional string rationale = 1;
  required string bucketId = 2;
}
```

```
message NodeOperationalStatusNoService {
   optional string rationale = 1;
   required string bucketId = 2;
}
```

```
NodeOperationalStatusNoServiceFrom {
  optional string rationale = 1;
  required string from = 2;  // date
  required string bucketId = 3;
}
```


```
NodeOperationalStatusNoServiceUntil {
  optional string rationale = 1;
  optional string from = 2  // date
  required string until = 3;  // date
  required string bucketId = 4;
}
```

Doing this I think will help address a couple issues raised in other places in the review as well as simplifying implementation in the commands and squid mappings I presume.

kdembler commented on query-node/mappings/src/storage/index.ts:
```diff
@@ -585,3 +599,89 @@ async function removeDistributionBucketOperator(store: DatabaseManager, operator
+
+  if ((await getWorker(store, workingGroup.name, workerId)) === undefined) {
+    return invalidMetadata(`The worker ${workerId} does not exist in the ${workingGroup.name} working group`)
+  }
```
I'm wondering, do we need to bother with worker for storage buckets? Those only have a single operator so I guess the worker ID doesn't make any difference?

kdembler commented on query-node/mappings/src/storage/metadata.ts:
```diff
@@ -66,6 +73,52 @@ async function processNodeLocationMetadata(
+    if (meta.noServiceFrom && meta.noServiceTo) {
+      const status = new NodeOperationalStatusNoServiceDuring()
+      status.from = new Date(meta.noServiceFrom)
+      status.to = new Date(meta.noServiceTo)
```
Are those raw dates validated anywhere? If not, we probably should do that

zeeshanakram3 replied:
Addressed this in storage squid PR https://github.com/Joystream/storage-squid/pull/19

kdembler commented on query-node/schemas/storageEvents.graphql:
```diff
@@ -0,0 +1,53 @@
+  ### SPECIFIC DATA ###
+
+  "Storage Bucket"
+  storageBucket: StorageBucket
```
Shouldn't that be a required field?

kdembler commented on storage-node/src/commands/leader/set-node-operational-status.ts:
```diff
@@ -0,0 +1,88 @@
+    }),
+    workerId: flags.integer({
+      char: 'w',
+      description: 'ID of the operator (distribution group worker)',
```
```suggestion
      description: 'ID of the operator (storage group worker)',
```

kdembler commented on storage-node/src/commands/operator/set-metadata.ts:
```diff
@@ -45,16 +57,55 @@ export default class OperatorSetMetadata extends ApiCommandBase {
     const encodedMetadata = '0x' + Buffer.from(StorageBucketOperatorMetadata.encode(metadata).finish()).toString('hex')
 
-    logger.info('Setting the storage operator metadata...')
+    logger.info('Se tting the storage node operational status...')
```
```suggestion
    logger.info('Setting the storage node operational status...')
```

kdembler replied:
Also, I think this still should be the old string since the command is also used for regular metadata?

kdembler commented on storage-node/src/commands/operator/set-metadata.ts:
```diff
@@ -45,16 +57,55 @@ export default class OperatorSetMetadata extends ApiCommandBase {
+              ? NodeOperationalStatusMetadata.OperationalStatus.NORMAL
+              : NodeOperationalStatusMetadata.OperationalStatus.NO_SERVICE,
+        },
+      }
```
I think if somebody will try to set metadata without operational status then this will set it to `NO_SERVICE`?

zeeshanakram3 replied:
Addressed in [6ff4531](https://github.com/Joystream/joystream/pull/4793/commits/6ff45319432ff711ce35e281f01b0fba889089a3)

kdembler commented on distributor-node/src/commands/operator/set-metadata.ts:
```diff
@@ -31,18 +43,58 @@ export default class OperatorSetMetadata extends AccountsCommandBase {
+            params.operationalStatus?.status === 'Normal'
+              ? NodeOperationalStatusMetadata.OperationalStatus.NORMAL
+              : NodeOperationalStatusMetadata.OperationalStatus.NO_SERVICE,
+        },
```
I think if somebody will try to set metadata without operational status then this will set it to `NO_SERVICE`?

zeeshanakram3 replied:
Addressed in [6ff4531](https://github.com/Joystream/joystream/pull/4793/commits/6ff45319432ff711ce35e281f01b0fba889089a3)

kdembler commented on storage-node/src/services/queryNode/schema.graphql:
```diff
@@ -957,6 +962,9 @@ type DistributionBucketOperatorMetadata implements BaseGraphQLObject {
   nodeLocationId: String
 
+  """Optional node operational status"""
+  nodeOperationalStatus: NodeOperationalStatus
```
I think this can be hard to query because of QN nested filtering limitations. I think we need to design a schema in a way that I can easily fetch all the active operators. If I'm not mistaken with the current schema the query would have to be (paraphrasing, I don't remember the query syntax for checking specific union type):
```gql
query {
  distributionBucketOperators(
    where: { metadata: { nodeOperationalStatus: { status: "OKAY" } } }
  ) {
    id
  }
}
```
And that filter would be too deep for QN to execute. Same applies to storage buckets

kdembler replied:
Also, would it be possible to make operational status required part of metadata? Because empty operational status currently would also mean "Okay" but it would be hard to query it I think

zeeshanakram3 replied:
I think you would not have that problem with storage squid's graphql server

kdembler reviewed (changes requested):
Great work @zeeshanakram3 and sorry for doing this review so late 🙇 I think one additional feature we could introduce is for distribution nodes to only query active storage operators when doing pings/asset checks. But we can do that as a followup later, this in itself is quite valuable.
Edit: I see that issue for that is already created and scoped: https://github.com/Joystream/joystream/issues/4816

zeeshanakram3 committed:
Merge remote-tracking branch 'upstream/master' into origo_argus_colossus_operational_status_feat

zeeshanakram3 committed:
address requested changes

zeeshanakram3 committed:
update generated storage-squid schema

zeeshanakram3 committed:
[Argus] ensure storage node only syncs from operationally active nodes

zeeshanakram3 committed:
update colossus docs

zeeshanakram3 committed:
[Argus] ensure distributor node only connects with operationally active storage nodes

zeeshanakram3 committed:
revert QN changes (i.e. remove mappings and schema changes)

zeeshanakram3 commented:
@kdembler I have addressed the requested changes. Also, removed the QN mappings needed for this feature as I think having the code at two places makes it difficult to review and maintain

Also, did changes as described in #4816 for both Argus/Colossus

zeeshanakram3 committed:
update graphql queries

kdembler commented on distributor-node/src/commands/leader/set-node-operational-status.ts:
```diff
@@ -0,0 +1,81 @@
+    const leadKey = await this.getDistributorLeadKey()
+
+    let operationalStatus: INodeOperationalStatusMetadata = {}
+    switch (statusType) {
```
`statusType` input is not required, should we default to `Normal`?

mnaamani replied:
It makes more sense to make it required, the command name doesn't imply what the default value should be.

zeeshanakram3 replied:
> It makes more sense to make it required, the command name doesn't imply what the default value should be.

Done

kdembler commented on distributor-node/src/commands/operator/set-metadata.ts:
```diff
@@ -31,18 +43,60 @@ export default class OperatorSetMetadata extends AccountsCommandBase {
     const workerKey = await this.getDistributorWorkerRoleKey(workerId)
 
+    let operationalStatus: INodeOperationalStatusMetadata = {}
+    switch (statusType) {
```
Same question about default status here

kdembler commented on distributor-node/src/commands/operator/set-metadata.ts:
```diff
@@ -31,18 +43,60 @@ export default class OperatorSetMetadata extends AccountsCommandBase {
+                  ? NodeOperationalStatusMetadata.OperationalStatus.NORMAL
+                  : NodeOperationalStatusMetadata.OperationalStatus.NO_SERVICE,
+            }
+          : {},
```
Would this override the current status? Or skip the update?

zeeshanakram3 replied:
It should skip the update but previously it was overriding, I have fixed it

kdembler commented on distributor-node/src/services/networking/query-node/queries/queries.graphql:
```diff
@@ -107,7 +107,17 @@ fragment StorageBucketOperatorFields on StorageBucket {
+      operatorMetadata: {
+        OR: [
+          { nodeOperationalStatus_isNull: true }
+          { nodeOperationalStatus: { isTypeOf_eq: "NodeOperationalStatusNormal" } }
```
This filter would skip any nodes with future maintenance as well. We probably should get all the operators and later only filter those that are under maintenance **currently**

zeeshanakram3 replied:
Done in [0537ee7](https://github.com/Joystream/joystream/pull/4793/commits/0537ee74c8f5caa8a14217668fc7cab59a5ca68c)

kdembler commented on storage-node/src/services/queryNode/queries/queries.graphql:
```diff
@@ -3,19 +3,62 @@ fragment StorageBucketIds on StorageBucket {
+          { nodeOperationalStatus_isNull: true }
+          { nodeOperationalStatus: { isTypeOf_eq: "NodeOperationalStatusNormal" } }
+        ]
+      }
```
Same comment here as with the query in distribution node

zeeshanakram3 replied:
Addressed

mnaamani commented on metadata-protobuf/proto/Storage.proto:
```diff
@@ -46,3 +48,31 @@ message DistributionBucketFamilyMetadata {
+
+message SetNodeOperationalStatus {
+  optional NodeOperationalStatusMetadata operational_status = 1; // Node's operational status to set
+  optional string worker_id = 2; // Storage/Distribution Worker ID
```
I would argue only the `bucket_id` is strictly necessary in the message.

zeeshanakram3 replied:
Correct for storage nodes, but for distributor nodes we also need to know about the `worker_id`, since a bucket can be assigned to multiple operators?

mnaamani commented on distributor-node/src/commands/leader/set-node-operational-status.ts:
```diff
@@ -0,0 +1,81 @@
+    bucketId: flags.bucketId({
+      required: true,
+    }),
+    workerId: flags.integer({
```
I suggested that worker id is not needed in metadata. If we agree on that, then this workerId can be dropped.

zeeshanakram3 replied:
Please see https://github.com/Joystream/joystream/pull/4793#discussion_r1538138524

mnaamani commented on distributor-node/src/command-base/default.ts:
```diff
@@ -98,6 +100,21 @@ export default abstract class DefaultCommandBase extends Command {
+        type: 'datepicker',
+        name: 'result',
+        clearable: true,
+        default: new Date('2017-09-28 17:36:05').toISOString(),
```
Tried the date picker, works really nicely in the terminal!

Perhaps a more reasonable default would be the current date `new Date().toISOString()`, so less fiddling with inital year and month for the user.

zeeshanakram3 replied:
Done

mnaamani reviewed (changes requested):
I'm suggesting to make multiple messages rather than a single one. Feels like a better design, but this will require a bit of refactoring here and in the storage-squid mappings.

zeeshanakram3 committed:
Merge remote-tracking branch 'upstream/master' into origo_argus_colossus_operational_status_feat

zeeshanakram3 committed:
address requested changes

zeeshanakram3 committed:
fix: only filter currently under maintenance nodes

zeeshanakram3 commented:
@mnaamani @kdembler I have addressed the requested changes. Also, re-designed the protobuf message schemas as requested. 

If the new protobuf messages are acceptable, then please publish the npm package so I would update the storage-squid mappings as well

mnaamani commented:
> @mnaamani @kdembler I have addressed the requested changes. Also, re-designed the protobuf message schemas as requested.
> 
> If the new protobuf messages are acceptable, then please publish the npm package so I would update the storage-squid mappings as well

published protobuf and joystream/js packages (for joystream/js I bumped the dependency on protobuf package)
